Modules/arraymodule.c

```diff
@@ -669,34 +805,176 @@ newarrayobject(PyTypeObject *type, Py_ssize_t size, const struct arraydescr *des
     return (PyObject *) op;
 }
 
+static inline int
+valid_index(Py_ssize_t i, Py_ssize_t limit)
+{
+    return (size_t) i < (size_t) limit;
+}
+
+static inline PyObject *
+getarrayitem(arrayobject *ap, Py_ssize_t i, char *items)
+{
+#ifndef NDEBUG
+    array_state *state = find_array_state_by_type(Py_TYPE(ap));
+    assert(array_Check(ap, state));
+#ifdef Py_GIL_DISABLED
+    assert(valid_index(i, (_Py_CONTAINER_OF(items, arraydata, items))->allocated));
+#else
+    assert(valid_index(i, Py_SIZE(ap)));
+#endif
+#endif
+    return (*ap->ob_descr->getitem)(items, i);
+}
+
+static PyObject *
+getarrayitem_locked(PyObject *op, Py_ssize_t i)
+{
+    PyObject *ret;
+    Py_BEGIN_CRITICAL_SECTION(op);
+#ifdef Py_GIL_DISABLED
+    if (!_PyObject_GC_IS_SHARED(op)) {
+        _PyObject_GC_SET_SHARED(op);
+    }
+#endif
+    if (!valid_index(i, Py_SIZE(op)))  {
+        ret = NULL;
+    }
+    else {
+        arrayobject *ap = (arrayobject *)op;
+        ret = getarrayitem(ap, i, ap->data->items);
+    }
+    Py_END_CRITICAL_SECTION();
+    return ret;
+}
+
+#ifdef Py_GIL_DISABLED
+
+static PyObject *
+getarrayitem_maybe_locked(PyObject *op, Py_ssize_t i)
```

Comment:
Let's name this `getarrayitem_threadsafe` to match the naming used in `dictobject.c`